### PR TITLE
Briefly discuss pros and cons of alternative designs

### PR DESCRIPTION
**Pull request by chris-wood**

### CODE REVIEW

File: draft-rescorla-tls-esni.md [hunk 1]
```diff
@@ -601,8 +601,52 @@ lie. In this design, the attacker would need to be able to find a
...
+application layer. In this section we describe several alternatives and discuss 
+drawbacks in comparison to the design in this document.
+
+## TLS-layer
+
+### TLS Tunnels
+
+In this variant, real TLS Client Hellos are tunneled within the early data payload 
+of an outer TLS connection established with the client-facing server. This requires 
+the client to have established a previous session — and obtained a PSK — with the
```
Comment (kazuho) [inline draft-rescorla-tls-esni.md]: I am not sure if "TLS tunnel" requires a previous session or the use of 0-RTT data.

Would you mind considering either adjusting the name or adjusting the text? Because what I expect for a scheme called "TLS tunnel" is as follows:

"TLS tunnel" is an approach that first establishes a TLS connection to a fronting host, and then, on top of that connection, establishes another TLS connection that uses the hostname of the hidden host.

If that is the case, I'd assume that all prior knowledge that is necessary is the binding between two names: name of the fronting host and that of the hidden host.

In other words, I'd expect that the "TLS Tunnel" scheme will be distributing using DNS the name of the fronting host instead of the ESNI key.

The downside of the tunneling is approach will be that there will be additional roundtrip when 0-RTT is not used, as well as the cost of double handshake and double encryption.

File: draft-rescorla-tls-esni.md [hunk 2]
```diff
@@ -601,8 +601,52 @@ lie. In this design, the attacker would need to be able to find a
...
+### HTTP/2 CERTIFICATE Frames 
+
+In this variant, clients request secondary certificates with CERTIFICATE_REQUEST HTTP/2 
+frames after TLS connection completion. In response, servers supply certificates via TLS 
+exported authenticators {{!I-D.ietf-tls-exported-authenticator}} in CERTIFICATE frames. 
+Clients use a generic SNI for the underlying client-facing server TLS connection. 
+Problems with this approach include: (1) one additional round trip before peer 
+authentication, and (2) non-trivial application-layer dependencies and interaction. 
+In contrast, encrypted SNI induces no additional round trip and operates below the 
+application layer.
```
Comment (kazuho) [inline draft-rescorla-tls-esni.md]: Doesn't the scheme also has the bootstrap problem: i.e. distributing the name of the client-facing server that corresponds to the name of the server that the client actually wants to connect to?

I can see that such binding can sometimes advertised eagerly using the CERTIFICATE frames, but I do not think that can be done for every host that the client might want to connect to.

Comment (chris-wood) [reply draft-rescorla-tls-esni.md]: Depending on how it's done, yeah, bootstrapping is also an issue here. I'll mention that.

File: draft-rescorla-tls-esni.md (hunk 1 above)
Comment (chris-wood) [reply draft-rescorla-tls-esni.md]: The tunnel approach you describe is a simple proxy. We can describe it here, though it's different from what was the "hide TLS in early data" proposal. I updated the title to try to make it more clear.

File: draft-rescorla-tls-esni.md [hunk 3]
```diff
@@ -606,8 +606,52 @@ lie. In this design, the attacker would need to be able to find a
...
+Alternative approaches to encrypted SNI may be implemented at the TLS or 
+application layer. In this section we describe several alternatives and discuss 
+drawbacks in comparison to the design in this document.
+
+## TLS-layer
+
+### TLS in Early Data
+
+In this variant, real TLS Client Hellos are tunneled within the early data payload 
+of an outer TLS connection established with the client-facing server. This requires
```
Comment (kazuho) [inline draft-rescorla-tls-esni.md]: Thank you for the changes.

Now I think I understand the approach suggested here; it's about encapsulating CH using 0-RTT data of an outer TLS connection, but sending the rest of the TLS transmissions directly (i.e. without double encryption).

Assuming that that is true, it might be worth clarifying that (e.g., state something like "Rest of the TLS messages / records are exchanged directly between the peers without double encryption").

Other than that, the PR looks to be ready for merge.

Comment (chris-wood) [reply draft-rescorla-tls-esni.md]: @kazuho I added some clarifying text. Please have a look!

File: draft-rescorla-tls-esni.md [hunk 4]
```diff
@@ -606,8 +606,57 @@ lie. In this design, the attacker would need to be able to find a
...
+
+### TLS in Early Data
+
+In this variant, TLS Client Hellos are tunneled within early data payloads
+belonging to outer TLS connections established with the client-facing server. This 
+requires clients to have established a previous session -— and obtained PSKs —- with 
+the server. The client-facing server decrypts early data payloads to uncover Client Hellos
+destined for the hidden server, and forwards them onwards as necessary. All Server
+Hello messages generated in response by hidden servers are forwarded directly to
+client -- unmodified. This avoids double encryption of TLS handshake messages.
```
Comment (kazuho) [inline draft-rescorla-tls-esni.md]: Thank you for explicitly stating that only Client Hello is sent as 0-RTT data.

Maybe we might want to change the sentence starting from "All Server Hello ..." to something like "Rest of the exchanges between the client and the hidden server, starting from Server Hello, are forwarded by the client-facing server -- unmodified.", to clarify that exchanges following Server Hello are also forwarded directly.

Other than that, the PR looks ready for merge to me.

Comment (kazuho) [reply draft-rescorla-tls-esni.md]: The reason I suggested the change is because the paragraph seemed to be unclear to me on whether or not the handshake messages and application data that follows Server Hello is double-encrypted, because it only discusses about Client Hello and Server Hello explicitly.

Comment (chris-wood) [reply draft-rescorla-tls-esni.md]: @kazuho Should be more clear now!

**Comment by kazuho** [comment]
@chris-wood Thank you for the changes. It's super clear to me now.
